include/boost/geometry/algorithms/simplify.hpp

```diff
@@ -116,34 +118,203 @@ struct simplify_range
+        // simplified away. If so, set start accordingly
+        Ring closing;
+        range::push_back(closing, range::at(ring, end - 1));
+        range::push_back(closing, range::at(ring, 0));
+        range::push_back(closing, range::at(ring, 1));
+
+        std::size_t simp_size = simplified_size(closing, distance, strategy);
+        while (simp_size == 2 && start < size - 2)
+        {
+            start++;
+
+            // Try to add more points at start:
+            range::push_back(closing, range::at(ring, start + 1));
+            simp_size = simplified_size(closing, distance, strategy);
+        }
+
+        // Verify if more points from end can be added
+        // Note that simplify works same in two directions.
+        // Avoid simplifying the whole geometry away,
+        // so avoid using the same point at end as at start,
+        geometry::clear(closing);
+        range::push_back(closing, range::at(ring, 0));
+        range::push_back(closing, range::at(ring, end - 1));
+        range::push_back(closing, range::at(ring, end - 2));
+
+        simp_size = simplified_size(closing, distance, strategy);
+        while (simp_size == 2 && end - 3 > start)
+        {
+            end--;
+
+            // Try to add more points from end:
+            range::push_back(closing, range::at(ring, end - 2));
+            simp_size = simplified_size(closing, distance, strategy);
+        }
+    }
+};
+
+
+struct simplify_ring
+{
+    template <typename Area>
+    static inline int area_sign(Area const& area)
+    {
+        return area > 0 ? 1 : area < 0 ? -1 : 0;
+    }
+
+    template <typename Ring, typename Strategy, typename Distance>
+    static inline void apply(Ring const& ring, Ring& out,
+                    Distance const& max_distance, Strategy const& strategy)
+    {
+        if (boost::empty(ring))
+        {
+            return;
+        }
+
+        static closure_selector const closure = geometry::closure<Ring>::value;
+
+        std::size_t const size = boost::size(ring);
+        std::size_t start = 0;
+        std::size_t end = size - 1; // index of last point
+
+        // Verify that what was positive, stays positive (or goes to 0)
+        // and what was negative stays negative (or goes to 0)
+        int const input_sign = area_sign(geometry::area(ring));
+
+        if (closure == geometry::closed)
+        {
+            // Verify area around closing point, if that can be simplified,
+            // start/end are modified and a corresponding slice will be used
+            // for simplification
+
+            // Take only a part of simplify distance, to avoid aggressive
+            // behaviour at closing points (it is known as a "open problem")
+
+            // TODO: for open polygons, implementation should be modified
+            simplify_closure_inspector::get_start_end(start, end, ring,
+                    max_distance / 2.0, strategy);
```

Comment:
Where does value 2.0 come from? Is it empirical (it was 10.0, 1000.0 etc in previous versions)? Is it a well known "open problem" or it comes from some testcase?

Reply:
The "open problem" is described here: http://cs.smith.edu/~jorourke/TOPP/P24.html#Problem.24
I did not look in more detail to those cited sources. But I understand the problem, you have to have a starting point (previously the closing point) and it is more or less an arbitrary starting point. If you rotate the polygon and do the simplification again, you might get something different.
This solution mitigates this problem by potentially simplifying the area around the closing point - but that is a different simplification, it still does not guarantee the optimal solution.

The 1000.0 earlier was actually hiding another bug in the first version (forgot to close the result), and I wanted to drastically reduce that later (10), but that is also arbitrary. And it did show some artefacts (but that can also be random). The value of 2.0 is still arbitrary too, it must be "some" value smaller than specified distance. But this did not give artefacts.

I will send you a picture later or maybe we can inlude them even in this review (Greece is one of the testcases!)

Reply:
But this is a problem of the strategy (i.e. Douglas-Peucker algorithm) that offer no guarantees for the maximum distance between the original and the simplified curve. Therefore, should this heuristic of decreasing `max_distance` be inside that strategy?